src/main/java/org/zalando/nakadi/repository/TopicRepositoryHolder.java

```diff
@@ -0,0 +1,102 @@
+
+import com.codahale.metrics.MetricRegistry;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+import org.zalando.nakadi.config.NakadiSettings;
+import org.zalando.nakadi.domain.Storage;
+import org.zalando.nakadi.exceptions.TopicRepositoryException;
+import org.zalando.nakadi.repository.kafka.KafkaFactory;
+import org.zalando.nakadi.repository.kafka.KafkaLocationManager;
+import org.zalando.nakadi.repository.kafka.KafkaSettings;
+import org.zalando.nakadi.repository.kafka.KafkaTopicRepository;
+import org.zalando.nakadi.repository.zookeeper.ZooKeeperHolder;
+import org.zalando.nakadi.repository.zookeeper.ZookeeperSettings;
+import org.zalando.nakadi.util.UUIDGenerator;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+@Component
+public class TopicRepositoryHolder {
+
+    private static final Logger LOG = LoggerFactory.getLogger(TopicRepositoryHolder.class);
+    private final NakadiSettings nakadiSettings;
+    private final KafkaSettings kafkaSettings;
+    private final ZookeeperSettings zookeeperSettings;
+    private final UUIDGenerator uuidGenerator;
+    private final MetricRegistry metricRegistry;
+    private final ReadWriteLock readWriteLock;
+    private final Map<Storage, TopicRepository> storageTopicRepository;
+
+    @Autowired
+    public TopicRepositoryHolder(final NakadiSettings nakadiSettings,
+                                 final KafkaSettings kafkaSettings,
+                                 final ZookeeperSettings zookeeperSettings,
+                                 final UUIDGenerator uuidGenerator,
+                                 final MetricRegistry metricRegistry) {
+        this.nakadiSettings = nakadiSettings;
+        this.kafkaSettings = kafkaSettings;
+        this.zookeeperSettings = zookeeperSettings;
+        this.uuidGenerator = uuidGenerator;
+        this.metricRegistry = metricRegistry;
+        this.readWriteLock = new ReentrantReadWriteLock();
+        this.storageTopicRepository = new HashMap<>();
+    }
+
+    public TopicRepository getTopicRepository(final Storage storage) throws TopicRepositoryException {
+        readWriteLock.readLock().lock();
+        try {
+            final TopicRepository topicRepository = storageTopicRepository.get(storage);
+            if (topicRepository != null) {
+                return topicRepository;
+            }
+        } finally {
+            readWriteLock.readLock().unlock();
+        }
+
+        readWriteLock.writeLock().lock();
+        try {
+            TopicRepository topicRepository = storageTopicRepository.get(storage);
+            if (topicRepository != null) {
+                return topicRepository;
+            }
+            topicRepository = createTopicRepository(storage);
```

Comment:
Topic repository holder is not under review in this PR (you can see it in description of PR, sorry about that, but I wanted to separate the work in favour of lighter reviews).
In my opinion it is fine that topic repository holder (maybe not the best name) creates repositories, but in the future it will know nothing about how to create the repository. You can see related changes in here https://github.com/zalando/nakadi/pull/565/files#diff-a161f5862153864c412c49e86a824adeR25